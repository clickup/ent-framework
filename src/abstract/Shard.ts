import type { Client } from "./Client";
import type { Query } from "./Query";
import type { QueryAnnotation, WhyClient } from "./QueryAnnotation";
import type { Timeline } from "./Timeline";

/**
 * Master freshness: reads always go to master.
 */
export const MASTER = Symbol("MASTER");

/**
 * Stale replica freshness: reads always go to a replica, even if it's stale.
 */
export const STALE_REPLICA = Symbol("STALE_REPLICA");

/**
 * Options passed to Shard constructor.
 */
export interface ShardOptions<TClient extends Client> {
  locateClient: (
    freshness: typeof MASTER | typeof STALE_REPLICA
  ) => Promise<TClient>;
  onRunError: (attempt: number, error: unknown) => Promise<"retry" | "throw">;
}

/**
 * Shard is a numbered island with one master and N replicas.
 */
export class Shard<TClient extends Client> {
  private shardClients = new WeakMap<TClient, TClient>();

  constructor(
    public readonly no: number,
    private options: ShardOptions<TClient>
  ) {}

  /**
   * Chooses the right client to be used for this shard. We don't memoize,
   * because the Shard may relocate to another Island during re-discovery.
   */
  async client(
    timeline: Timeline | typeof MASTER | typeof STALE_REPLICA
  ): Promise<TClient> {
    const { client } = await this.clientEx(timeline, undefined);
    return client;
  }

  /**
   * Runs a query after choosing the right client (destination connection,
   * shard, annotation etc.)
   */
  async run<TOutput>(
    query: Query<TOutput>,
    annotation: QueryAnnotation,
    timeline: Timeline,
    freshness: null | typeof MASTER | typeof STALE_REPLICA
  ): Promise<TOutput> {
    for (let attempt = 0; ; attempt++) {
      const { client, whyClient } = await this.clientEx(
        freshness ?? timeline,
        query.IS_WRITE ? true : undefined
      );

      let res;
      try {
        res = await query.run(client, {
          ...annotation,
          whyClient,
          attempt: annotation.attempt + attempt,
        });
      } catch (error: unknown) {
        // The shard was there by the moment we got its client, but it probably
        // disappeared (during migration) and appeared on some other island.
        if ((await this.options.onRunError(attempt, error)) === "retry") {
          continue;
        } else {
          throw error;
        }
      }

      if (query.IS_WRITE && freshness !== STALE_REPLICA) {
        timeline.setPos(
          await client.timelineManager.currentPos(),
          client.timelineManager.maxLagMs
        );
      }

      return res;
    }
  }

  /**
   * Throws if this shard does not exist, or its island is down, or something
   * else is wrong with it.
   */
  async assertDiscoverable(): Promise<void> {
    await this.options.locateClient(MASTER);
  }

  /**
   * An extended client selection logic. There are multiple reasons (8+ in total
   * so far) why a master or a replica may be chosen to send the query to. We
   * don't memoize, because the Shard may relocate to another Island during
   * re-discovery.
   */
  private async clientEx(
    timeline: Timeline | typeof MASTER | typeof STALE_REPLICA,
    isWrite: true | undefined
  ): Promise<{ client: TClient; whyClient: WhyClient }> {
    if (isWrite) {
      return {
        client: await this.shardClient(MASTER),
        whyClient: "master-bc-is-write",
      };
    }

    if (timeline === MASTER) {
      return {
        client: await this.shardClient(MASTER),
        whyClient: "master-bc-master-freshness",
      };
    }

    const replica = await this.shardClient(STALE_REPLICA);

    if (replica.isMaster) {
      return {
        client: replica,
        whyClient: "master-bc-no-replicas",
      };
    }

    if (timeline === STALE_REPLICA) {
      return {
        client: replica,
        whyClient: "replica-bc-stale-replica-freshness",
      };
    }

    const isCaughtUp = timeline.isCaughtUp(
      await replica.timelineManager.currentPos()
    );
    return isCaughtUp
      ? {
          client: replica,
          whyClient: isCaughtUp,
        }
      : {
          client: await this.shardClient(MASTER),
          whyClient: "master-bc-replica-not-caught-up",
        };
  }

  /**
   * Returns a shard-aware Client of a particular freshness.
   */
  private async shardClient(
    freshness: typeof MASTER | typeof STALE_REPLICA
  ): Promise<TClient> {
    const client = await this.options.locateClient(freshness);
    let shardClient = this.shardClients.get(client);
    if (!shardClient) {
      shardClient = client.withShard(this.no);
      this.shardClients.set(client, shardClient);
    }

    return shardClient;
  }
}
