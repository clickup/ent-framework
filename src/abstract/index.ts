/**
 * @file Automatically generated by barrelsby.
 */

export * from "./Batcher";
export * from "./Client";
export * from "./ClientError";
export * from "./Cluster";
export * from "./Island";
export * from "./Loader";
export * from "./LocalCache";
export * from "./Loggers";
export * from "./Query";
export * from "./QueryAnnotation";
export * from "./QueryBase";
export * from "./QueryPing";
export * from "./Runner";
export * from "./Schema";
export * from "./Shard";
export * from "./ShardError";
export * from "./ShardIsNotDiscoverableError";
export * from "./ShardNamer";
export * from "./Timeline";
export * from "./TimelineManager";
