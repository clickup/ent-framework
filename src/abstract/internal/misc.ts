export const OP_TIMELINE_POS_REFRESH = "TIMELINE_POS_REFRESH";
export const OP_PING = "PING";
export const OP_SHARD_NOS = "SHARD_NOS";
