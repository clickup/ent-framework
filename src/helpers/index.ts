/**
 * @file Automatically generated by barrelsby.
 */

export * from "./isBigintStr";
export * from "./testSpecTypeIntegrity";
