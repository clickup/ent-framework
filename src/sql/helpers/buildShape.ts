const RE_COMMENT = toRegExp`
  /\*.*?\*/
`;

const RE_NUMBER = toRegExp`
  (?<!\w)\d+
`;

const RE_STRING = toRegExp`
  ' (?:[^']|'')* '
`;

const RE_IN_ROW = toRegExp`
  (
    \sIN\(
    ROW \( [^)]+ \)
  )
  (?: ,\s* ROW \( [^)]+ \) )*
`;

const RE_WITH_VALUES = toRegExp`
  ^(
    WITH \s [^\n]+VALUES\n
      [^\n]+,\n
      ( \s* \( (?:[^\n]+) \) )
  ),\n
    (?: \2 (?:,\n)? )+
`;

const RE_INSERT_VALUES = toRegExp`
  ^(
    INSERT \s INTO \s [^\n]+VALUES\n
      ( \s* \( (?:[^\n]+) \) )
  ),\n
    (?: \2 (?:,\n)? )+
`;

const RE_IDENTICAL_UNION_ALL = toRegExp`
  (?<=^|\n)
  ( \(? SELECT \s .*? )\n
  (?:
    (\s* UNION \s ALL)\n
    \1 \n?
  )+
`;

export default function buildShape(sql: string) {
  return sql
    .trim()
    .replace(RE_COMMENT, "")
    .replace(RE_NUMBER, "?")
    .replace(RE_STRING, "'?'")
    .replace(RE_IN_ROW, "$1, ...")
    .replace(RE_WITH_VALUES, "$1, ...")
    .replace(RE_INSERT_VALUES, "$1, ...")
    .replace(RE_IDENTICAL_UNION_ALL, "$1\n$2 ...");
}

function toRegExp(template: TemplateStringsArray) {
  return new RegExp(template.raw.join("").replace(/\s+/g, ""), "g");
}
