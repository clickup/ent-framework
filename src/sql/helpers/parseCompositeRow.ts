/**
 * Parses composite row input into parts. See references at:
 * - https://www.postgresql.org/docs/current/rowtypes.html
 * - https://github.com/DmitryKoterov/db_type/blob/master/lib/DB/Type/Pgsql/Row.php
 * - unit tests of this function
 */
export default function parseCompositeRow(str: string): Array<string | null> {
  let p = 0;
  let c = "";

  function readCharAfterSpaces(): string {
    while (p < str.length) {
      const c = str[p];
      if (c !== " " && c !== "\t" && c !== "\r" && c !== "\n") {
        break;
      }
    }

    return p < str.length ? str[p] : "";
  }

  function throwError(message: string) {
    throw Error(`${message} at position ${p} of "${str}"`);
  }

  // Leading "(".
  c = readCharAfterSpaces();
  if (c !== "(") {
    throwError('An anonymous composite type row must start with "("');
  }

  p++;

  // Check for immediate trailing ')': by convention, "()" translates to
  // ROW(NULL) tuple and not to an empty tuple (because ROW() and ROW(NULL) have
  // the same stringified representation which is "()" unfortunately).
  c = readCharAfterSpaces();
  if (c === ")") {
    return [null];
  }

  // Row may contain:
  // - "-quoted strings (escaping: ["] is doubled)
  // - unquoted strings (before first "," or ")")
  // - empty string (it is treated as NULL)
  // Nested rows and all other things are represented as strings.

  const reTillNext = /[^,)]*/sy;
  const reQuoted = /"((?:[^"]+|"")*)"/sy;

  const result: Array<string | null> = [];
  while (true) {
    // We read a value in this iteration, then - delimiter.
    c = readCharAfterSpaces();

    // Always read a next element value.
    if (c === "," || c === ")") {
      // Comma or end of row instead of value: treat as NULL.
      result.push(null);
    } else if (c !== '"') {
      // Unquoted string. Notice that NULL here is treated as "NULL" string, but
      // NOT as a null value! This is how composite values are encoded by PG.
      reTillNext.lastIndex = p;
      const matches = reTillNext.exec(str)!;
      result.push(matches[0]);
      p += matches[0].length;
    } else {
      reQuoted.lastIndex = p;
      const matches = reQuoted.exec(str);
      if (matches) {
        // Quoted string.
        result.push(matches[1].replace('""', '"').replace("\\\\", "\\"));
        p += matches[0].length;
      } else {
        throwError("Expected a balanced quoted string");
      }
    }

    // Delimiter or the end of row.
    c = readCharAfterSpaces();
    if (c === ",") {
      p++;
      continue;
    } else if (c === ")") {
      p++;
      break;
    } else {
      throwError('Expected a delimiter "," or ")"');
    }
  }

  return result;
}
