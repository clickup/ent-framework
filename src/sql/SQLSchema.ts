import type { Query } from "../abstract/Query";
import { Schema } from "../abstract/Schema";
import type {
  CountInput,
  ExistsInput,
  InsertInput,
  LoadByInput,
  Row,
  SelectByInput,
  SelectInput,
  Table,
  UniqueKey,
  UpdateInput,
} from "../types";
import { SQLQueryCount } from "./SQLQueryCount";
import { SQLQueryDelete } from "./SQLQueryDelete";
import { SQLQueryExists } from "./SQLQueryExists";
import { SQLQueryIDGen } from "./SQLQueryIDGen";
import { SQLQueryInsert } from "./SQLQueryInsert";
import { SQLQueryLoad } from "./SQLQueryLoad";
import { SQLQueryLoadBy } from "./SQLQueryLoadBy";
import { SQLQuerySelect } from "./SQLQuerySelect";
import { SQLQuerySelectBy } from "./SQLQuerySelectBy";
import { SQLQueryUpdate } from "./SQLQueryUpdate";
import { SQLQueryUpsert } from "./SQLQueryUpsert";

export class SQLSchema<
  TTable extends Table,
  TUniqueKey extends UniqueKey<TTable>
> extends Schema<TTable, TUniqueKey> {
  idGen(): Query<string> {
    return new SQLQueryIDGen(this);
  }

  insert(input: InsertInput<TTable>): Query<string | null> {
    return new SQLQueryInsert(this, input);
  }

  upsert(input: InsertInput<TTable>): Query<string> {
    return new SQLQueryUpsert(this, input);
  }

  update(id: string, input: UpdateInput<TTable>): Query<boolean> {
    return new SQLQueryUpdate(this, id, input);
  }

  delete(id: string): Query<boolean> {
    return new SQLQueryDelete(this, id);
  }

  load(id: string): Query<Row<TTable> | null> {
    return new SQLQueryLoad(this, id);
  }

  loadBy(input: LoadByInput<TTable, TUniqueKey>): Query<Row<TTable> | null> {
    return new SQLQueryLoadBy(this, input);
  }

  selectBy(
    input: SelectByInput<TTable, TUniqueKey>
  ): Query<Array<Row<TTable>>> {
    return new SQLQuerySelectBy(this, input);
  }

  select(input: SelectInput<TTable>): Query<Array<Row<TTable>>> {
    return new SQLQuerySelect(this, input);
  }

  count(input: CountInput<TTable>): Query<number> {
    return new SQLQueryCount(this, input);
  }

  exists(input: ExistsInput<TTable>): Query<boolean> {
    return new SQLQueryExists(this, input);
  }
}
