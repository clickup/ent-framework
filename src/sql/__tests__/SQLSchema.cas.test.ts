import assert from "assert";
import type { Shard } from "../../abstract/Shard";
import { MASTER } from "../../abstract/Shard";
import { join, nullthrows } from "../../helpers/misc";
import { SQLSchema } from "../SQLSchema";
import type { TestSQLClient } from "./test-utils";
import {
  ByteaBuffer,
  EncryptedValue,
  TEST_TIMELINE,
  recreateTestTables,
  shardRun,
  testCluster,
} from "./test-utils";

const schema = new SQLSchema(
  'sql-schema.cas"table',
  {
    name: { type: String },
    ts: { type: Date, allowNull: true },
    encrypted_field: {
      type: EncryptedValue,
      allowNull: true,
      autoInsert: "NULL",
    },
    buffer_field: {
      type: ByteaBuffer,
      allowNull: true,
      autoInsert: "NULL",
    },
    created_at: { type: Date, autoInsert: "now()" },
    updated_at: { type: Date, autoUpdate: "now()" },
    id: { type: String, autoInsert: "id_gen()" },
  },
  ["name"]
);

let shard: Shard<TestSQLClient>;
let master: TestSQLClient;
let replica: TestSQLClient;

beforeEach(async () => {
  await recreateTestTables([
    {
      CREATE: [
        `CREATE TABLE %T(
          id bigint NOT NULL PRIMARY KEY,
          name text NOT NULL,
          ts timestamptz,
          encrypted_field text,
          buffer_field bytea,
          created_at timestamptz NOT NULL,
          updated_at timestamptz NOT NULL,
          UNIQUE (name)
        )`,
      ],
      SCHEMA: schema,
      SHARD_AFFINITY: [],
    },
  ]);

  shard = await testCluster.randomShard();
  master = await shard.client(MASTER);
  master.resetSnapshot();
  replica = await shard.client(TEST_TIMELINE);
  replica.resetSnapshot();
});

test("CAS single success", async () => {
  const ts1 = new Date("2000-01-01T00:00:00.001Z");
  const ts1upd = new Date("2000-02-01T00:00:00.001Z");

  const [id1, id2] = await join([
    shardRun(
      shard,
      schema.insert({
        name: "a",
        ts: ts1,
        encrypted_field: await EncryptedValue.encrypt("a", 42),
        buffer_field: Buffer.from([1, 2, 3]),
      })
    ),
    shardRun(shard, schema.insert({ name: "b", ts: null })),
  ]);
  assert(id1 && id2);
  const [row1, row2] = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  assert(row1 && row2);
  master.resetSnapshot();

  const r1 = await shardRun(
    shard,
    schema.update(nullthrows(id1), {
      name: "a-upd",
      ts: ts1upd,
      $cas: {
        name: "a",
        ts: ts1,
        encrypted_field: await EncryptedValue.encrypt("a", 42),
        buffer_field: Buffer.from([1, 2, 3]),
        updated_at: row1.updated_at,
      },
    })
  );
  const r2 = await shardRun(
    shard,
    schema.update(nullthrows(id2), {
      name: "b-upd",
      ts: row2.updated_at,
      $cas: { name: "b", ts: null, updated_at: row2.updated_at },
    })
  );
  master.toMatchSnapshot();

  expect(r1).toBeTruthy();
  expect(r2).toBeTruthy();

  const rows = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  expect(rows).toMatchObject([
    { id: id1, name: "a-upd", ts: ts1upd },
    { id: id2, name: "b-upd", ts: row2.updated_at },
  ]);
});

test("CAS single skip", async () => {
  const [id1, id2] = await join([
    shardRun(shard, schema.insert({ name: "a", ts: new Date() })),
    shardRun(
      shard,
      schema.insert({
        name: "b",
        ts: null,
        buffer_field: Buffer.from([1, 2, 3]),
      })
    ),
  ]);
  assert(id1 && id2);
  const [row1, row2] = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  assert(row1 && row2);
  master.resetSnapshot();

  const r1Skip = await shardRun(
    shard,
    schema.update(nullthrows(id1), {
      name: "a-skip",
      $cas: { name: "a-old", updated_at: row1.updated_at },
    })
  );
  const r2Skip = await shardRun(
    shard,
    schema.update(nullthrows(id2), {
      name: "b-skip",
      $cas: { buffer_field: Buffer.from([4, 5, 6]) },
    })
  );
  master.toMatchSnapshot();

  expect(r1Skip).toBeFalsy();
  expect(r2Skip).toBeFalsy();

  const rows = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  expect(rows).toMatchObject([
    { id: id1, name: "a", ts: row1.ts, updated_at: row1.updated_at },
    { id: id2, name: "b", ts: null, updated_at: row2.updated_at },
  ]);
});

test("CAS batched success", async () => {
  const ts1 = new Date("2000-01-01T00:00:00.001Z");
  const ts1upd = new Date("2000-02-01T00:00:00.001Z");

  const [id1, id2] = await join([
    shardRun(shard, schema.insert({ name: "a", ts: ts1 })),
    shardRun(shard, schema.insert({ name: "b", ts: null })),
  ]);
  assert(id1 && id2);
  const [row1, row2] = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  assert(row1 && row2);
  master.resetSnapshot();

  const [r1, r2] = await join([
    shardRun(
      shard,
      schema.update(nullthrows(id1), {
        name: "a-upd",
        ts: ts1upd,
        $cas: { name: "a", ts: ts1, updated_at: row1.updated_at },
      })
    ),
    shardRun(
      shard,
      schema.update(nullthrows(id2), {
        name: "b-upd",
        ts: row2.updated_at,
        $cas: { name: "b", ts: null, updated_at: row2.updated_at },
      })
    ),
  ]);
  master.toMatchSnapshot();

  expect(r1).toBeTruthy();
  expect(r2).toBeTruthy();

  const rows = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  expect(rows).toMatchObject([
    { id: id1, name: "a-upd", ts: ts1upd },
    { id: id2, name: "b-upd", ts: row2.updated_at },
  ]);
});

test("CAS batched success and skip", async () => {
  const [id1, id2] = await join([
    shardRun(shard, schema.insert({ name: "a", ts: new Date() })),
    shardRun(shard, schema.insert({ name: "b", ts: null })),
  ]);
  assert(id1 && id2);
  const [row1, row2] = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  assert(row1 && row2);

  const [r1, r2] = await join([
    shardRun(
      shard,
      schema.update(nullthrows(id1), {
        name: "a-upd",
        $cas: { name: "a", updated_at: row1.updated_at },
      })
    ),
    shardRun(
      shard,
      schema.update(nullthrows(id2), {
        name: "b-upd",
        $cas: { name: "b-old", updated_at: row2.updated_at },
      })
    ),
  ]);
  expect(r1).toBeTruthy();
  expect(r2).toBeFalsy();

  const rows = await join([
    shardRun(shard, schema.load(id1)),
    shardRun(shard, schema.load(id2)),
  ]);
  expect(rows).toMatchObject([
    { id: id1, name: "a-upd" },
    { id: id2, name: "b", updated_at: row2.updated_at },
  ]);
});

test("CAS batched success and skip for the same row", async () => {
  const id1 = nullthrows(
    await shardRun(shard, schema.insert({ name: "a", ts: new Date() }))
  );
  const row1 = nullthrows(await shardRun(shard, schema.load(id1)));
  master.resetSnapshot();

  const [r1, r2] = await join([
    shardRun(
      shard,
      schema.update(nullthrows(id1), {
        name: "a-upd",
        $cas: { name: "a", updated_at: row1.updated_at },
      })
    ),
    shardRun(
      shard,
      schema.update(nullthrows(id1), {
        name: "a-skip",
        $cas: { name: "a", updated_at: new Date(42) },
      })
    ),
  ]);
  master.toMatchSnapshot();

  expect(r1).toBeTruthy();
  expect(r2).toBeFalsy();

  const rows = await join([shardRun(shard, schema.load(id1))]);
  expect(rows).toMatchObject([{ id: id1, name: "a-upd" }]);
});
