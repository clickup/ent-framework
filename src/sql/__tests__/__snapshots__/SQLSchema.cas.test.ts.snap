// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CAS batched success 1`] = `
"
WITH rows(id, name, ts, updated_at, "$cas.name", "$cas.ts", "$cas.updated_at", _key) AS (VALUES
  ((NULL::"sql-schema.cas""table").id, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").ts, (NULL::"sql-schema.cas""table").updated_at, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").ts, (NULL::"sql-schema.cas""table").updated_at, '<key>'),
  ('<id>', 'a-upd', '<date>', now(), 'a', '<date>', '<date>', '<id>:<hash>')<reordered for test>,
  ('<id>', 'b-upd', '<date>', now(), 'b', NULL, '<date>', '<id>:<hash>')<reordered for test>)
  UPDATE "sql-schema.cas""table" SET name=rows.name, ts=rows.ts, updated_at=rows.updated_at
  FROM rows WHERE "sql-schema.cas""table".id=rows.id AND ROW("sql-schema.cas""table".name, date_trunc('ms', "sql-schema.cas""table".ts), date_trunc('ms', "sql-schema.cas""table".updated_at)) IS NOT DISTINCT FROM ROW(rows."$cas.name", rows."$cas.ts", rows."$cas.updated_at") RETURNING rows._key
----
WITH rows(id, name, ts, updated_at, "$cas.name", "$cas.ts", "$cas.updated_at", _key) AS (VALUES
  ((NULL::"sql-schema.cas""table").id, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").ts, (NULL::"sql-schema.cas""table").updated_at, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").ts, (NULL::"sql-schema.cas""table").updated_at, '?'),
  ('?', '?', '?', now(), '?', '?', '?', '?')<reordered for test>,
  ('?', '?', '?', now(), '?', NULL, '?', '?')<reordered for test>)
  UPDATE "sql-schema.cas""table" SET name=rows.name, ts=rows.ts, updated_at=rows.updated_at
  FROM rows WHERE "sql-schema.cas""table".id=rows.id AND ROW("sql-schema.cas""table".name, date_trunc('?', "sql-schema.cas""table".ts), date_trunc('?', "sql-schema.cas""table".updated_at)) IS NOT DISTINCT FROM ROW(rows."$cas.name", rows."$cas.ts", rows."$cas.updated_at") RETURNING rows._key
"
`;

exports[`CAS batched success and skip for the same row 1`] = `
"
WITH rows(id, name, updated_at, "$cas.name", "$cas.updated_at", _key) AS (VALUES
  ((NULL::"sql-schema.cas""table").id, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").updated_at, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").updated_at, '<key>'),
  ('<id>', 'a-skip', now(), 'a', '<date>', '<id>:<hash>')<reordered for test>,
  ('<id>', 'a-upd', now(), 'a', '<date>', '<id>:<hash>')<reordered for test>)
  UPDATE "sql-schema.cas""table" SET name=rows.name, updated_at=rows.updated_at
  FROM rows WHERE "sql-schema.cas""table".id=rows.id AND ROW("sql-schema.cas""table".name, date_trunc('ms', "sql-schema.cas""table".updated_at)) IS NOT DISTINCT FROM ROW(rows."$cas.name", rows."$cas.updated_at") RETURNING rows._key
----
WITH rows(id, name, updated_at, "$cas.name", "$cas.updated_at", _key) AS (VALUES
  ((NULL::"sql-schema.cas""table").id, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").updated_at, (NULL::"sql-schema.cas""table").name, (NULL::"sql-schema.cas""table").updated_at, '?'),
  ('?', '?', now(), '?', '?', '?'), ...<reordered for test>)
  UPDATE "sql-schema.cas""table" SET name=rows.name, updated_at=rows.updated_at
  FROM rows WHERE "sql-schema.cas""table".id=rows.id AND ROW("sql-schema.cas""table".name, date_trunc('?', "sql-schema.cas""table".updated_at)) IS NOT DISTINCT FROM ROW(rows."$cas.name", rows."$cas.updated_at") RETURNING rows._key
"
`;

exports[`CAS single skip 1`] = `
"
UPDATE "sql-schema.cas""table" SET name='a-skip', updated_at=now() WHERE id='<id>' AND ROW(name, date_trunc('ms', updated_at)) IS NOT DISTINCT FROM ROW('a-old', '<date>') RETURNING id
----
UPDATE "sql-schema.cas""table" SET name='?', updated_at=now() WHERE id='?' AND ROW(name, date_trunc('?', updated_at)) IS NOT DISTINCT FROM ROW('?', '?') RETURNING id

UPDATE "sql-schema.cas""table" SET name='b-skip', updated_at=now() WHERE id='<id>' AND ROW(buffer_field) IS NOT DISTINCT FROM ROW('\\x040506') RETURNING id
----
UPDATE "sql-schema.cas""table" SET name='?', updated_at=now() WHERE id='?' AND ROW(buffer_field) IS NOT DISTINCT FROM ROW('?') RETURNING id
"
`;

exports[`CAS single success 1`] = `
"
UPDATE "sql-schema.cas""table" SET name='a-upd', ts='<date>', updated_at=now() WHERE id='<id>' AND ROW(name, date_trunc('ms', ts), encrypted_field, buffer_field, date_trunc('ms', updated_at)) IS NOT DISTINCT FROM ROW('a', '<date>', 'encrypted:Â‹', '\\x010203', '<date>') RETURNING id
----
UPDATE "sql-schema.cas""table" SET name='?', ts='?', updated_at=now() WHERE id='?' AND ROW(name, date_trunc('?', ts), encrypted_field, buffer_field, date_trunc('?', updated_at)) IS NOT DISTINCT FROM ROW('?', '?', '?', '?', '?') RETURNING id

UPDATE "sql-schema.cas""table" SET name='b-upd', ts='<date>', updated_at=now() WHERE id='<id>' AND ROW(name, date_trunc('ms', ts), date_trunc('ms', updated_at)) IS NOT DISTINCT FROM ROW('b', NULL, '<date>') RETURNING id
----
UPDATE "sql-schema.cas""table" SET name='?', ts='?', updated_at=now() WHERE id='?' AND ROW(name, date_trunc('?', ts), date_trunc('?', updated_at)) IS NOT DISTINCT FROM ROW('?', NULL, '?') RETURNING id
"
`;
